### PR TITLE
Revist logging

### PR DESCRIPTION
**Pull request by nojaf**
Currently, a lot of logging happens directly to the standard output. 
This is unwanted in the `Fable.Compiler`, and I wish to introduce a logging abstraction.
Inside my use case, I would like to log in to memory or disk.

The only thing I'm not quite sure about is what to do with https://github.com/fable-compiler/Fable/issues/2727. It seems like some code was written to avoid this problem and I'm not sure it will still be problematic after this change.
@AngelMunoz, is this still an active use-case of yours?

### CODE REVIEW

**Comment by MangelMaxime** [comment]
Reading the linked issue in your command, ringed a bell to me.

I believe we still have some issues when running Fable in watch mode.

I don't know if this is related to the logger or not but, when running a sub process via Fable sometimes the processes are freezing.

Example of command: `dotnet fable --run npx vite`

https://github.com/fable-compiler/Fable/issues/3631

I observed a similar behaviour when working on fantomas-tool or even worth sometimes for that particular project where `vite` process would start and just stop completely at some point.

**Comment by AngelMunoz** [comment]
@nojaf The main problem there was that the logs were being re-written in the same line via cursor manipulation, this could cause hang ups if the output was redirected (e.g. calling Fable via Process APIs or in my case CLIWrap).

From a quick glance on this PR, looks like the main change is that logging now goes through an `ILogger` right?

I'm guessing that leaves using the cursor to write in the same line/position out of the window, if that is the case, I believe it should not break anyone calling fable from another process via stdio.

**Comment by nojaf** [comment]
Ok, thanks for confirming. For now, I think I'd like to remove that cursor manipulation thing, as it is probably tricky to achieve using the ILogger.

**Comment by MangelMaxime** [comment]
The cursor manipulation allows for a much smaller / easier to read output in the log when Fable does all the recompilation of the files. So it would be nice to keep it.

There is currently the `--verbose` flag which I believe allow people to disable it if it needed.

**Comment by nojaf** [comment]
Ok, I restored that cursor update functionality.
Ready for review.

**Comment by MangelMaxime** [comment]
I don't have much time this week, review will most likely happen next weeks.

**Comment by ncave** [comment]
@nojaf Minor CI error to fix:
```
Error: /home/runner/work/Fable/Fable/tests/Integration/Compiler/Util/Compiler.fs(36,9): error FS0764: No assignment given for field 'Verbosity' of type 'Fable.Compiler.Util.CliArgs'
```

**Comment by nojaf** [comment]
> Minor CI error to fix:

Thanks, is this project not part of the `Fable.sln`?

**Comment by ncave** [comment]
@nojaf 
> is this project not part of the Fable.sln?

I think [it is](https://github.com/fable-compiler/Fable/blob/main/Fable.sln#L47).

**Comment by nojaf** [comment]
Alright, another crucial part of `Fable.Compiler` is that **no code writes to stdout**! 
This messed up the `JSON RPC` communication I'm using for the Vite plugin.

To guarantee that nobody ever sins against this, I've created an analyzer that detects calls to `prinfn` and `Console.Write`. The results are logged [here](https://github.com/fable-compiler/Fable/security/code-scanning?query=pr%3A3671+is%3Aopen+severity%3Aerror). And it is possible to configure a PR build to fail if errors are present. (Needs to be configured in GitHub).

`build / analyzers` is green because the process of running the analyzers succeeded. This won't fail if the analyzers detect something bad.

`Code scanning results / Ionide.Analyzers.Cli` is green because no **new** code violated the rules. In future PRs, we want this to be red when someone breaks the rules.

I still need to address the actual `Console.Write` usage in `Fable.Transform`, so please don't merge this PR yet. No need to convert to draft, I should be able to address this in the next days.

File: src/Fable.Analyzers/StdOutAnalyzer.fs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+open FSharp.Analyzers.SDK
+
+let namesToAvoid =
+    set
+        [
+            "System.Console.Write"
+            "System.Console.WriteLine"
+            "System.Console.WriteAsync"
+            "System.Console.WriteLineAsync"
+            "Microsoft.FSharp.Core.Operators.print"
```
Comment (ncave) [inline src/Fable.Analyzers/StdOutAnalyzer.fs:16]: @nojaf Shouldn't the last 3 be these instead?
```fs
    "Microsoft.FSharp.Core.Printf.PrintFormat"
    "Microsoft.FSharp.Core.Printf.PrintFormatToError"
    "Microsoft.FSharp.Core.Printf.PrintFormatLine"
    "Microsoft.FSharp.Core.Printf.PrintFormatLineToError"
```

Comment (nojaf) [reply src/Fable.Analyzers/StdOutAnalyzer.fs:16]: Thanks, those names are wrong indeed: https://github.com/dotnet/fsharp/blob/main/src/FSharp.Core/fslib-extra-pervasives.fs#L267-267

I'll check them.

**Comment by ncave** [comment]
@nojaf
> I still need to address the actual Console.Write usage in Fable.Transform

I moved the `Console.WriteLine` usage out of `Fable.Transforms`, please rebase.
You can now move the `Log` module back to `Fable.Compiler`.
This should fix your `Fable.Transforms` build issue.

**Comment by nojaf** [comment]
@ncave thanks! I've rebased and all is green now.

@MangelMaxime with the `getPlugin` changes in here I'd like to have this PR merged.
I was able to use my local Vite plugin for a small Fable project I have.
It would be great to have another `Fable.Compiler` release with this.

**Comment by MangelMaxime** [comment]
> @MangelMaxime with the `getPlugin` changes in here I'd like to have this PR merged.

Noted, I will work on the next release of Fable.

> It would be great to have another `Fable.Compiler` release with this.

To trigger the release for `Fable.Compiler` can you please update its Changelog file? You can update the `Unreleased` section for now, I will bump the version when doing the release process.

If you could update the `Unreleased` section of the different Changelog files impacted by this PR it would be great  🙏

**Comment by nojaf** [comment]
Done!

File: src/Fable.Transforms/Fable.Transforms.fsproj [hunk 2]
```diff
@@ -45,6 +45,9 @@
     <Compile Include="Rust/Fable2Rust.fs" />
     <Compile Include="State.fs" />
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="8.0.0" />
```
Comment (ncave) [inline src/Fable.Transforms/Fable.Transforms.fsproj:49]: @nojaf Perhaps this can be removed.

File: tests/Js/Adaptive/Fable.Tests.Adaptive.fsproj [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>net6.0</TargetFramework>
```
Comment (ncave) [inline tests/Js/Adaptive/Fable.Tests.Adaptive.fsproj:3]: @nojaf Please change to
```
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
```
for consistency with the other test projects.

File: src/Fable.Analyzers/Fable.Analyzers.fsproj [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
```
Comment (ncave) [inline src/Fable.Analyzers/Fable.Analyzers.fsproj:4]: @nojaf Can this target `net8.0` if it only run in CI?

**Comment by MangelMaxime** [comment]
Thank you for the contribution and to the reviewers

**Pull request merged**